runtime/src/storable_accounts.rs

```diff
@@ -0,0 +1,38 @@
+//! trait for abstracting underlying storage of pubkey and account pairs to be written
+use solana_sdk::{account::ReadableAccount, clock::Slot, pubkey::Pubkey};
+
+/// abstract access to pubkey, account, slot, target_slot of either:
+/// a. (slot, &[&Pubkey, &ReadableAccount])
+/// b. (slot, &[&Pubkey, &ReadableAccount, Slot]) (we will use this later)
+/// This trait avoids having to allocate redundant data when there is a duplicated slot parameter.
+/// All legacy callers do not have a unique slot per account to store.
+pub trait StorableAccounts<'a, T: ReadableAccount + Sync>: Sync {
+    /// pubkey at 'index'
+    fn pubkey(&self, index: usize) -> &Pubkey;
+    /// account at 'index'
+    fn account(&self, index: usize) -> &T;
+    /// slot that all accounts are to be written to
+    fn target_slot(&self) -> Slot;
+    /// true if no accounts to write
+    fn is_empty(&self) -> bool;
+    /// # accounts to write
+    fn len(&self) -> usize;
+}
```

Comment:
Did a previous version have a `fn slot()` too? Did you scrap that since we can always do `self.0`?

Reply:
yes, I removed slot() because nobody calls it yet. In the follow-on change to this to allow per-account 'slot', this method will be added and used.

Reply:
I don't think a caller can call `slot.0`. The caller only knows they have a `StorableAccounts` object. Only these trait methods are available on it.